### PR TITLE
Mean Reversion Portfolio Construction Model

### LINKED ISSUE #6518: No implementation on Mean-Reverting Portfolio
#### Expected Behavior
Mean-Reverting Portfolio (MRP) is a portfolio constrcution/optimization technique that has been used in the quant industry for decades. It is expected that LEAN would have a Portfolio Construction Model on that.

#### Actual Behavior
No abstraction of any kind of MRP tool is available in LEAN.

#### Potential Solution
Implementation of On-Line MRP in form of Portfolio Construction Model.

#### Checklist

- [X] I have completely filled out this template
- [X] I have confirmed that this issue exists on the current `master` branch
- [X] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by LouisSzeto**
#### Description
Implementation of Mean-Reversion Portfolio Construction Model, in form of On-Line Moving Average Reversion (OLMAR) Portfolio.

#### Related Issue
closes #6518

#### Requires Documentation Change
Writing Algorithm \ Algorithm Framework \ Portfolio Construction \ Supported Models

#### How Has This Been Tested?
Basic implementation is already done on `PortfolioConstructionModel` test. Added unit test on calculating the correct quantity of order.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                i => { },
+                true,
+                new DataPermissionManager()));
+        }
+
+        [TestCase(Language.CSharp, null, null, 47, 47)]
+        [TestCase(Language.Python, null, null, 47, 47)]
+        [TestCase(Language.CSharp, 1, -0.5, 31, 63)]
+        [TestCase(Language.Python, 1, -0.5, 31, 63)]
+        public void CorrectWeightings(Language language, double? magnitude1, double? magnitude2, decimal expectedQty1, decimal expectedQty2)
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs:68]: Cool! but think we should add more tests, there are a few cases that would be nice to cover, different arguments `epc,resolution,windowSize`, also see other PCMs test suites 👍

File: Algorithm.Framework/Portfolio/MeanReversionPortfolioConstructionModel.cs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            var targets = new Dictionary<Insight, double>();
+
+            // If we have no insights or non-ready just return an empty target list
+            if (activeInsights.IsNullOrEmpty() || 
+                !activeInsights.All(x => _symbolData[x.Symbol].IsReady()))
+            {
+                return targets;
+            }
+
+            var m = activeInsights.Count();
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Portfolio/MeanReversionPortfolioConstructionModel.cs:72]: `Count()` is using linq, can use Count to get the list property directly

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Cool! Leaving a few requests 👍

File: Algorithm.Framework/Portfolio/MeanReversionPortfolioConstructionModel.cs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+using Accord.Math;
+using QuantConnect.Algorithm.Framework.Alphas;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Indicators;
+using QuantConnect.Util;
+
+namespace QuantConnect.Algorithm.Framework.Portfolio
+{
+    /// <summary>
+    /// Implementation of On-Line Moving Average Reversion (OLMAR)
```
[the remark this answers is not in the document]
Comment (LouisSzeto) [reply Algorithm.Framework/Portfolio/MeanReversionPortfolioConstructionModel.cs:28]: Nope. This is correct. It was just the implementation MRP name

File: Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs (hunk 1 above)
Comment (LouisSzeto) [reply Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs]: I've updated that to include edge cases and keep the theoretical cases suggested by the original author's paper. They should be a good check for the model to work properly. Yet, I dun think toning the hyperparameters and copying the backtested values into it is appropriate.

File: Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs [hunk 4]
```diff
@@ -95,28 +142,97 @@ public virtual void SetUp()
...
+
+        [TestCase(Language.CSharp, 1)]
+        [TestCase(Language.Python, 1)]
+        [TestCase(Language.CSharp, 0.5)]
+        [TestCase(Language.Python, 0.5)]
+        [TestCase(Language.CSharp, 0)]
+        [TestCase(Language.Python, 0)]
+        public void SimplexProjection(Language language, double regulator)
+        {
+            var model = GetPortfolioConstructionModel(language, PortfolioBias.Long, Resolution.Daily);
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs:213]: `model` isn't used really, could call the python version for python language too I believe. Same for `GetPriceRelativesWithInsightMagnitude` & `GetPriceRelatives`

**Review by Martin-Molinero — APPROVED** [review]
